core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -567,11 +567,31 @@ object Doc {
    */
   private[paiges] case class Align(doc: Doc) extends Doc
 
-  private[paiges] class Thunk(value: => Doc) extends Serializable {
-    lazy val evaluated: Doc = value
-  }
+  private[paiges] case class LazyDoc(thunk: () => Doc) extends Doc {
+    private var computed: Doc = null
+    // This is never a LazyDoc
+    lazy val evaluated: Doc = {
+      @tailrec
+      def loop(d: Doc, toSet: List[LazyDoc]): Doc =
```

Comment:
I found the names here a bit confusing. I didn't notice the shadowing of `thunk` at first and thought it was a bug. Also, somehow the name `toSet` threw me off a bit. I think we should match `lzy@LazyDoc(f)` and call the list `toUpdate` instead.